### ISSUE TITLE
Provide better error message for 32bit vs 64bit missmatch.

### ISSUE THREAD

**Issue by treeform** [nim-lang/Nim#22352]
### Summary

When using 64bit Nim with 32bit C nothing but confusion happens.

### Description

For some reason or another using a 64bit Nim with 32bit GCC happens way too often. The error message it produces is very confusing. Nim should be able to error or generate the C code that errors (through #error directive) and explains what happens. Some thing like: 

"Nim's 64bit mode used with C's 32bit mode. Please install the right compiler or pass the right flags."

It seems like everyone has run into this and always has been confused by it:

* https://forum.nim-lang.org/t/10363
* https://forum.nim-lang.org/t/7059#68721
* https://forum.nim-lang.org/t/9821#64782
* https://forum.nim-lang.org/t/9643#63373

### Alternatives

* The C code can have an #error directive instead of the assert.
* The Nim compiler can warn about this even before generating the C code.
* Nim can generate 64/32 bit independed code instead.

### Examples

```
nim c -r foo64.nim 
```

Output:

```
...
...
...
Nim's 64bit mode used with C's 32bit mode. Please install the right compiler or pass the right flags.
```

**Comment by demotomohiro** [comment]
https://github.com/nim-lang/Nim/pull/19549 added error message to pointer size check assert and it was merged to Nim 2.0.

You can see how that error message looks like on console with following code:
```nim
{.emit: """
#define NIM_INTBITS 123
NIM_STATIC_ASSERT(sizeof(NI) == sizeof(void*) && NIM_INTBITS == sizeof(NI)*8, "Pointer size mismatch between Nim and C/C++ backend. You probably need to setup the backend compiler for target CPU.");
"""}
```

> It seems like everyone has run into this and always has been confused by it:
> 
> * https://forum.nim-lang.org/t/10363
> * https://forum.nim-lang.org/t/7059#68721
> * https://forum.nim-lang.org/t/9821#64782
> * https://forum.nim-lang.org/t/9643#63373

These threads were created before Nim 2.0 and that `NIM_STATIC_ASSERT` didn't have error message in Nim 1.6.14 and older.

> * The C code can have an #error directive instead of the assert.

`#error` directive is a part of C preprocessor and it is used with `#if` branch.
But it cannot evaluate `sizeof` operator.
So following code result in compile error.

```c
#define NIM_INTBITS 64
#if sizeof(NI) == sizeof(void*) && NIM_INTBITS == sizeof(NI)*8
#else
#error "Some error message"
#endif
```

**Comment by treeform** [comment]
Thank you for adding this. Hopefully it will help some people!
